### PR TITLE
make maxMajorProtVer easier to correctly maintain

### PR DESCRIPTION
**Pull request by nfrisby**
See the message of the single commit in this PR.

### CODE REVIEW

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 1]
```diff
@@ -579,7 +579,18 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
     -- The major protocol version of the last era is the maximum major protocol
     -- version we support.
     maxMajorProtVer :: MaxMajorProtVer
-    maxMajorProtVer = MaxMajorProtVer (pvMajor protVerBabbage)
+    maxMajorProtVer =
+        nonEmptyLast (exactlyWeakenNonEmpty maxMajorProtVers)
+      where
+        protVers :: Exactly (CardanoShelleyEras StandardCrypto) MaxMajorProtVer
+        protVers = Exactly $
+          -- ensure that these have the same order as 'CardanoShelleyEras'!
```
Comment (amesgen) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:587]: As discussed, a more invasive approach that does not have this (very minor) potential for future incorrect updates would be to modify `ProtocolParams*` to be type-family based and take an `NP` of those as an argument to `protocolInfoCardano`, which the last protocol version can then be derived from.

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 2]
```diff
@@ -579,7 +579,18 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
...
+          protVerShelley :*
+          protVerAllegra :*
+          protVerMary :*
+          protVerAlonzo :*
+          protVerBabbage :*
+          Nil
```
Comment (amesgen) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:593]: This fails to compile; this should work:
```suggestion
          MaxMajorProtVer
        . pvMajor
        . nonEmptyLast
        . exactlyWeakenNonEmpty
        $ protVers
      where
        protVers :: Exactly (CardanoShelleyEras StandardCrypto) ProtVer
        protVers = Exactly $
          -- ensure that these have the same order as 'CardanoShelleyEras'!
             K protVerShelley
          :* K protVerAllegra
          :* K protVerMary
          :* K protVerAlonzo
          :* K protVerBabbage
          :* Nil
```

**Review by amesgen — CHANGES_REQUESTED** [review]
Nice, this will make it very unlikely that we forget to update this in the future!

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:591]: I very much agree. Let's merge this PR now, and open a TechDebt for the follow-up (as it involves at least some `cardano-node` changes etc)

**Review by amesgen — APPROVED** [review]

**Pull request merged**